### ISSUE TITLE
After some time docker stops running containers with error 'HTTP error: Error response from daemon: couldn't get vxlan id for \"10.0.58.0/24\": failed to allocate vxlan id: no bit available" status=500' 

### ISSUE THREAD

**Issue by skryzhny** [moby/moby#29756]
**Description**

We have several docker 1.11.1 nodes in `swarm version 1.2.3 (eaa53c7)` standalone cluster. We using external Key-Value storage for overlay networking. We tried ETCD 2.3.1 and 2.5.1 then Zookeeper latest.

We are constantly creating overlay networks and running containers with that networks attached, then removing containers and networks.
Our setup is running good for days, but after some period docker start throwing errors "couldn't get vxlan id for \"10.0.58.0/24\": failed to allocate vxlan id: no bit available"
After that we can create new networks and run containers only after removing some existing networks and containers. And shortly we can't create networks/run containers even if there is no active networks/containers in swarm cluster.
We used to stop swarm/ docker on nodes/ KV storage, clear KV storage and then start all for curing that problem.

**Steps to reproduce the issue:**
We can't reliable reproduce that on staging env.
Partially reproduced issue by moving KV storage's data when errors occur from prod to staging.
After 1.5 hours of running jmeter tests it starts failing on staging too.

**Describe the results you received:**
```
# docker -H localhost:23750 network create -d overlay ssss3
2fc7f3e7346092a150592fe246eb84fcfd3e4e9508de41618a34c50acb033498
# docker -H localhost:23750 run -ti --network=ssss3 busybox sh
docker: Error response from daemon: Error response from daemon: couldn't get vxlan id for "10.0.8.0/24": failed to allocate vxlan id: no bit available.
# docker -H localhost:23750 network inspect ssss3
[
    {
        "Name": "ssss3",
        "Id": "2fc7f3e7346092a150592fe246eb84fcfd3e4e9508de41618a34c50acb033498",
        "Scope": "global",
        "Driver": "overlay",
        "EnableIPv6": false,
        "IPAM": {
            "Driver": "default",
            "Options": {},
            "Config": [
                {
                    "Subnet": "10.0.8.0/24",
                    "Gateway": "10.0.8.1/24"
                }
            ]
        },
        "Internal": false,
        "Containers": {},
        "Options": {},
        "Labels": {}
    }
]
```
**Describe the results you expected:**
New containers are able to run.

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

```
Client:
 Version:      1.11.1
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   5604cbe
 Built:        Wed Apr 27 00:34:42 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.11.1
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   5604cbe
 Built:        Wed Apr 27 00:34:42 2016
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 4
 Running: 3
 Paused: 1
 Stopped: 0
Images: 48
Server Version: 1.11.1
Storage Driver: devicemapper
 Pool Name: docker-202:0-786672-pool
 Pool Blocksize: 65.54 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: xfs
 Data file: /dev/vg-docker/data
 Metadata file: /dev/vg-docker/metadata
 Data Space Used: 106.4 GB
 Data Space Total: 288.8 GB
 Data Space Available: 182.4 GB
 Metadata Space Used: 169.8 MB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 1.978 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Library Version: 1.02.107-RHEL7 (2016-06-09)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: host overlay bridge null
Kernel Version: 3.10.0-327.36.3.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 29.55 GiB
Name: node7.********.***
ID: H6AE:UQRN:YZ5X:IHHI:ZJCC:BWNK:66YA:2Y5K:7WDY:7YUP:FUGC:VG4W
Docker Root Dir: /var/lib/docker
Debug mode (client): false
Debug mode (server): false
Registry: https://index.docker.io/v1/
WARNING: bridge-nf-call-ip6tables is disabled
Cluster store: zk://*********.***:2181
Cluster advertise: 172.30.12.107:2375
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
AWS EC2 r3.xlarge 
OS: Centos 7.2
Docker storage: LVM-Direct

Please tell that info/logs can help investigate.
Also I filled question here: https://forums.docker.com/t/what-resource-is-used-up-when-we-get-error-failed-to-allocate-vxlan-id-no-bit-available/26397

**Comment by cpuguy83** [comment]
ping @sanimej

**Comment by TylerJewell** [comment]
This issue is proving to be one that is destabilizing our system. Generally, our system is supporting 100s of concurrent containers with large memory footprints running across a variety of hardware nodes, sharing a key-value storage. Generally, containers are not long-running, but they are not being created / destroyed at the same rate as a CI-system might.

We are working to make this more reproducible, and have been able to replicate it a few times in our staging environment with load from JMeter. We would appreciate Docker looking into this, as we think it's potentially a serious bug. Happy for us to run additional diagnostics where possible.

**Comment by aboch** [comment]
Do you see any interesting warn or error daemon logs ?

**Comment by aboch** [comment]
@skryzhny 
Do you see any interesting warn or error daemon logs ?

When you hit the issue, please send me the data you get from
`GET http://<store IP>:8500/v1/kv/docker/network/v1.0/idm/vxlan-id/`
if you are using consul or the equivalent for the datastore you are using.

**Comment by skryzhny** [comment]
I haven't seen any unusual errors or warning in swarm log.

I have full copy of KV database from last malfunction.
It taken after stopping all services which deletes all containers and networks.
```
$ zookeepercli --servers localhost -c get /docker/network/v1.0/idm/vxlan-id
{"id":"vxlan-id","sequence":"AAAAAAAAAukAAAAAAAAAIv////8AAAAAAAAABf//+/8AAAAAAAAAAf//9/8AAAAAAAAAAb//7/8AAAAAAAAAAZ73//4AAAAAAAAAAf//v38AAAAAAAAAAf//O/IAAAAAAAAAAf//tz8AAAAAAAAAAe3n//8AAAAAAAAAAf////8AAAAAAAAAA/77//8AAAAAAAAAAf////8AAAAAAAAAAf7///8AAAAAAAAAAf////wAAAAAAAAAAf//9/8AAAAAAAAAAf////8AAAAAAAAAAf3///sAAAAAAAAAAf2AAAAAAAAAAAAAAQ=="}
```

And data from staging:
```
# zookeepercli --servers localhost -c get /docker/network/v1.0/idm/vxlan-id
{"id":"vxlan-id","sequence":"AAAAAAAAAukAAAAAAAAAAP////8AAAAAAAAAF/+AAAAAAAAAAAAAAQ=="}
```

**Comment by aboch** [comment]
Thanks @skryzhny 

After decoding the first data structure you provided, I see that out of the 745 ids overlay driver allocates for the vxlan networks, there are still 34 available to use.

I tested requesting an extra one from the sequence, and the first available was returned and its reservation was done properly in the backend data structure, which is a bitmask. I tested releasing some, and the release was done correctly as well.

The vxlan id is requested at the time the overlay network is plumbed, which is when the first container on that network is run/connected. The release happens when the network is removed.

Now that I decoded the second structure, the one from your staging state, I in fact see that all the 745 ids are taken.

Clearly the vxlan id release on overlay network deletion either is not taking place or some more subtle bug, given you see nothing suspicious in the logs.

I will try to reproduce on 1.11.1 with some loops of overlay nw create/docker run/network rm. Is there any other details I should know about the setup ? Are you specifying more than one network segment for each overlay network (multiple `--subnet`) ?

Thanks

**Comment by sanimej** [comment]
@skryzhny Are you using compose files ? If you can share the exact sequence of container/network create and destroy that might give some clues.

**Comment by skryzhny** [comment]
@aboch 
First data structure should have only 15-20 ids taken, as `docker -H localhost:23750 network ls` reported that amount of networks. Probably not deleted correctly by some reasons like timeouts or our product crash because of unrelated reasons,
At that time we had demand for about 60 networks. We already had some (34?) allocated but can't allocate more.
So I used to stop all services. Then our product recycled remaining active containers and networks, so that 34 ids released.

We are constantly creating and removing overlay networks at high rate like 1000 or more per day.
Number of successfully created and deleted networks clearly more than 8000
Also we run several times our jmeter test with 2500 iterations each without hitting the problem.
So most of the times we are able to recycle ids.

Second structure was produced on staging from first one.
Our staging have upgraded docker 1.2.5.
Jmeter test use 10 threads, so no more than 10 networks should be active at any given time.
It took about hour to start failing.

I think we using only one network segment per network. @garagatyi can you confirm it?

No, we use our own product that call docker API via swarm standalone.
We have implemented some subset of compose syntax and functionality.

Can you share tool or how-to determine current amount of free/used ids?

**Comment by skryzhny** [comment]
@sanimej: I can increase swarm/docker loglevel to debug on staging if You want it.

We use KV storage only for docker, so I think I might find a way to obfuscate sensitive information and share it.

**Comment by aboch** [comment]
@skryzhny 

> Can you share tool or how-to determine current amount of free/used ids?

Sure, I wrote this to decode it:
https://gist.github.com/aboch/76bea69c38a1b90ac3b13ea477fc131f

go build it and run it like this:

```
./bitseq native '{"id":"vxlan-id","sequence":"AAAAAAAAAukAAAAAAAAAAP////8AAAAAAAAAF/+AAAAAAAAAAAAAAQ=="}'
```
or in case the ds gives you the value in bytes (like from the consul ui):
```
./bitseq Value '{"CreateIndex":23,"ModifyIndex":4647,"LockIndex":0,"Key":"docker/network/v1.0/idm/vxlan-id/","Flags":0,"Value":"eyJpZCI6InZ4bGFuLWlkIiwic2VxdWVuY2UiOiJBQUFBQUFBQUF1a0FBQUFBQUFBQzZRQUFBQUFBQUFBQUFBQUFHQT09In0="}'
```

The output will contain the number of bits, the number of available bits and the bits allocation mask as a run-length encoded sequence where each symbol is a 32 bit long bitmask represented in hex.

So this one, for example:
```
./bitseq Value '{"CreateIndex":23,"ModifyIndex":6464,"LockIndex":0,"Key":"docker/network/v1.0/idm/vxlan-id/","Flags":0,"Value":"eyJpZCI6InZ4bGFuLWlkIiwic2VxdWVuY2UiOiJBQUFBQUFBQUF1a0FBQUFBQUFBQ3VmLy8vLzhBQUFBQUFBQUFBUUFBLy84QUFBQUFBQUFBQVFBQUFBQUFBQUFBQUFBQUZnPT0ifQ=="}'
Bit sequence:
App: , ID: vxlan-id, DBIndex: 0x0, bits: 745, unselected: 697, sequence: (0xffffffff, 1)->(0xffff, 1)->(0x0, 22)->end
```
says that 48 out of the 745 bits are allocated. And the allocation is distributed as:  
One block of 32 bits fully allocated, followed by one block with the _last_ 16 bits allocated, followed by 22 blocks of 32 non allocated bits. 
Read each element in the list as: `(32 bit bitmask symbol, number of equal consecutive symbols)`

To generate this example, I had created 64 overlay networks with running container, then disconnected and removed the networks from 33 to 48. That's why the 16 bits hole, otherwise it would look like `(0xffffffff,2)->(0x0, 22)->end`.

The vxlan id mgr is built on top of this. So bit 0 => vxlan 256 and so on.

Hope this will help you debugging your issue.

**Comment by garagatyi** [comment]
We use single segment per network. Actually we use default segment settings for now.

**Comment by skryzhny** [comment]
@aboch Thank for utility, I'll add it to our monitoring system.

```
# docker -H localhost:23750 network ls | grep -c [o]verlay
19
# get-bits.sh
{"id":"vxlan-id","sequence":"AAAAAAAAAukAAAAAAAAA6JQftSMAAAAAAAAAAX////8AAAAAAAAAAf////8AAAAAAAAAAf7/97cAAAAAAAAAAXbf//8AAAAAAAAAAf//d/8AAAAAAAAAAf////8AAAAAAAAAAv//7/0AAAAAAAAAAf/t728AAAAAAAAAAW/eOv0AAAAAAAAAAbY7//8AAAAAAAAAAf////8AAAAAAAAAAf/9//0AAAAAAAAAAf//3/8AAAAAAAAAAf39828AAAAAAAAAAf9fVhoAAAAAAAAAAd////8AAAAAAAAAAd8AAAAAAAAAAAAAAQAAAAAAAAAAAAAABQ=="}
Bit sequence: App: , ID: vxlan-id, DBIndex: 0x0, bits: 745, unselected: 232, sequence: (0x941fb523, 1)->(0x7fffffff, 1)->(0xffffffff, 1)->(0xfefff7b7, 1)->(0x76dfffff, 1)->(0xffff77ff, 1)->(0xffffffff, 2)->(0xffffeffd, 1)->(0xffedef6f, 1)->(0x6fde3afd, 1)->(0xb63bffff, 1)->(0xffffffff, 1)->(0xfffdfffd, 1)->(0xffffdfff, 1)->(0xfdfdf36f, 1)->(0xff5f561a, 1)->(0xdfffffff, 1)->(0xdf000000, 1)->(0x0, 5)->end
```
Looks like we really have problems here, allocated more than 500 bits for 19 networks

**Comment by aboch** [comment]
@skryzhny 
I tried reproducing with docker 1.11.2 (no relevant changes in this area compared to 1.11.1) using zookeeper (zk running as a container on a separate VM and with a script decoding the sequence retrieved via zookeepercli), with 200 overlay networks, disconnecting and removing networks with different patterns, then recreating some and removing others: https://gist.github.com/aboch/0cc880b80e272dce86c91d4a5476f417

But I was not able to hit anything wrong in the allocation/release of the vxlan ids.

There must be something else in your use case that we are missing.

Is it possible for you to retrieve the overlay network creation and removal pattern ?
Something like "created 25 networks, removed 1st, 13th, created 5 networks, removed ..." so that I can replay the same events manually and see if the logic is failing because of that specific pattern.
I would like to rule out any bug in this area first.

Thanks

**Comment by garagatyi** [comment]
I'll try to explain what if happening in our system from docker POV.
We have users. Each of them may create some containerized environment. 
These environments are created in parallel. Amount of parallel threads is not limited, but not very high for now.
All operations are performed over Docker Swarm  (old one). 
Also Swarm is NOT connected to KV at all, e.g. uses file for nodes list lookup.
Creation of each environment:
 - create docker network
POST /v1.20/networks/create
```
{
  "Name": "workspacefq7l9ldhwper36d0_nt6rk6ouc0hqlgv6",
  "CheckDuplicate": true,
  "Driver": "overlay",
  "Internal": false,
  "EnableIPv6": false
}
```
 - create container (May be repeated several times and depends on user's environment definition). I cut JSON to remove not relevant fields and sensitive information. If you need other fields, please inform me.
POST /v1.20/containers/create?name=workspacefq7l9ldhwper36d0_machinexdvp2855selnhunq
```
{
  "Image": "eclipse-che/workspacefq7l9ldhwper36d0_machinexdvp2855selnhunq_admin_dev-machine",
  "NetworkDisabled": false,
  "HostConfig": {
    "Links": [],
    "PublishAllPorts": true,
    "Privileged": false,
    "ExtraHosts": [
      "che-host:172.18.0.11"
    ],
    "VolumesFrom": [],
    "NetworkMode": "workspacefq7l9ldhwper36d0_nt6rk6ouc0hqlgv6",
  },
  "NetworkingConfig": {
    "EndpointsConfig": {
      "workspacefq7l9ldhwper36d0_nt6rk6ouc0hqlgv6": {
        "Links": [],
        "Aliases": [
          "dev-machine"
        ]
      }
    }
  }
}
```
 - start container
 - reading container logs and other operations with a specific containers (not related to networks).

After some time environment gets destroyed either by user or by system due to timeout
 - remove container with force (for each container in user's environment)
 - remove image that was source of container
 - remove network

**Comment by skryzhny** [comment]
We have the same behaviour on our another test environment, nightly.
If you can't reproduce the problem, I think we have next options:
- I may share docker or/and swarm logs with debug turned on, but it's a lot of data to analyse.
- I may share JMeter test script, docs how  to setup environment like we have on AWS.
- I may give root ssh access to node with docker and JMeter script we used.

**Comment by skryzhny** [comment]
We moved to docker 1.12.5 and bits limit increased from 745 to 16776960 so that problem not such critical for us at the moment.

We still have that problem and after two days of running we got
`Free 16776130 from 16776960`

**Comment by thaJeztah** [comment]
ping @aboch what's the status here?

**Comment by thaJeztah** [comment]
Let me close this ticket for now, as it looks like it went stale.
